### PR TITLE
Add missing documentation in cudf/types.hpp

### PR DESCRIPTION
**Pull request by karthikeyann**
Fixes parts of https://github.com/rapidsai/cudf/issues/9373
added missing documentation in types.hpp to fix doxygen warnings
fixes 30 warnings

### CODE REVIEW

File: cpp/include/cudf/types.hpp [hunk 1]
```diff
@@ -32,17 +32,20 @@
  * @brief Type declarations for libcudf.
  */
 
+// FIXME: where is this used? TODO: remove if unused.
 namespace bit_mask {
 using bit_mask_t = uint32_t;
 }
```
Comment (karthikeyann) [inline cpp/include/cudf/types.hpp:39]: `bit_mask_t` is unused.
Can we remove this?

File: cpp/include/cudf/types.hpp [hunk 2]
```diff
@@ -32,17 +32,20 @@
...
 
 // Forward declarations
+/// @cond FORWARD_DECLARATIONS
 namespace rmm {
 class device_buffer;
 namespace mr {
 class device_memory_resource;
 device_memory_resource* get_current_device_resource();
 }  // namespace mr
+/// @endcond
```
Comment (karthikeyann) [inline cpp/include/cudf/types.hpp:48]: Using `@cond` to skip this forward declarations by doxygen.

File: cpp/include/cudf/types.hpp (hunk 1 above)
Comment (davidwendt) [reply cpp/include/cudf/types.hpp:39]: Yes. We have `bitmask_type` defined in this same file and that is what is being used.

File: cpp/include/cudf/types.hpp (hunk 2 above)
Comment (davidwendt) [reply cpp/include/cudf/types.hpp:48]: Cool trick.

File: cpp/include/cudf/types.hpp [hunk 3]
```diff
@@ -243,9 +246,21 @@ class data_type {
  public:
   data_type()                 = default;
   ~data_type()                = default;
-  data_type(data_type const&) = default;
-  data_type(data_type&&)      = default;
+  data_type(data_type const&) = default;  ///< Copy constructor
+  data_type(data_type&&)      = default;  ///<  Move constructor
```
Comment (davidwendt) [inline cpp/include/cudf/types.hpp:250]:
```suggestion
  data_type(data_type&&)      = default;  ///< Move constructor
```

File: cpp/include/cudf/types.hpp (hunk 2 above)
Comment (vyasr) [reply cpp/include/cudf/types.hpp:48]: This is interesting. When and how would we actually enable `FORWARD_DECLARATIONS`?

**Review by vyasr — APPROVED** [review]
LGTM

File: cpp/include/cudf/types.hpp (hunk 2 above)
Comment (davidwendt) [reply cpp/include/cudf/types.hpp:48]: I would think we would never enable it -- so doxygen skips this part -- if I'm understanding this.
https://www.doxygen.nl/manual/commands.html#cmdcond
It appears the label is optional so we could remove if that is less confusing.

**Comment by karthikeyann** [comment]
rerun tests

**Review by davidwendt — APPROVED** [review]

**Pull request merged**